### PR TITLE
feat: abstract out majority of bootstrap CLI code

### PR DESCRIPTION
**Pull request by jaredallard**
## What this PR does / why we need it

This PR pulls most of the CLI-specific code we have out of bootstrap and instead, putting it inside of gobox. This makes the code much easier to maintain / test and is less duplicated.

Example PR on the devenv: https://github.com/getoutreach/devenv/pull/130

**JIRA ID**: [DTSS-0]

### CODE REVIEW

File: pkg/cli/bootstrap.go [hunk 1]
```diff
@@ -0,0 +1,255 @@
...
+	"github.com/getoutreach/gobox/pkg/log"
+	"github.com/getoutreach/gobox/pkg/secrets"
+	"github.com/getoutreach/gobox/pkg/trace"
+	"github.com/getoutreach/gobox/pkg/updater"
+	"github.com/sirupsen/logrus"
+	"github.com/urfave/cli/v2"
+	"gopkg.in/yaml.v2"
+)
+
+const (
```
Comment (george-e-shaw-iv) [inline pkg/cli/bootstrap.go:31]: const block needs a comment for lintroller platinum, even tho this isn't platinum we might as well get with the program

File: pkg/cli/bootstrap.go [hunk 2]
```diff
@@ -0,0 +1,255 @@
...
+	var fallbackSecretLookup func(context.Context, string) ([]byte, error)
+	fallbackSecretLookup = secrets.SetDevLookup(func(ctx context.Context, key string) ([]byte, error) {
+		if key == "APIKey" {
+			return []byte(honeycombApiKey), nil
+		}
+
+		return fallbackSecretLookup(ctx, key)
+	})
+
+	log.SetOutput(io.Discard)
```
Comment (george-e-shaw-iv) [inline pkg/cli/bootstrap.go:48]: is this just to make sure we're using logrus over gobox log

File: pkg/cli/bootstrap.go [hunk 3]
```diff
@@ -0,0 +1,255 @@
...
+
+// urfaveRegisterShutdownHandler registers a signal notifier that translates various term
+// signals into context cancel
+func urfaveRegisterShutdownHandler(cancel context.CancelFunc) {
+	// handle ^C gracefully
+	c := make(chan os.Signal, 1)
+	signal.Notify(c, os.Interrupt, syscall.SIGTERM, syscall.SIGHUP)
+	go func() {
+		<-c
+		cancel()
```
Comment (george-e-shaw-iv) [inline pkg/cli/bootstrap.go:93]:
```suggestion
		<-c
		signal.Reset()
		cancel()
```

File: pkg/cli/bootstrap.go [hunk 4]
```diff
@@ -0,0 +1,255 @@
...
+	}()
+}
+
+// setupTracer sets up a root trace for the CLI and initializes the tracer
+func setupTracer(ctx context.Context, name string) context.Context {
```
Comment (george-e-shaw-iv) [inline pkg/cli/bootstrap.go:98]: maybe modify this function to return a closer function that calls `trace.CloseTracer(ctx)`

Comment (george-e-shaw-iv) [reply pkg/cli/bootstrap.go:98]: nvm i saw what you did. kinda sketchy but i understand where it comes from with how generic this code has to be

File: pkg/cli/doc.go [hunk 5]
```diff
@@ -0,0 +1,3 @@
+// Package cli contains various cli utilities that are useful for building
+// cli applications with gobox.
+package cli
```
Comment (george-e-shaw-iv) [inline pkg/cli/doc.go:3]:
```suggestion
// Copyright 2021 Outreach Corporation. All Rights Reserved.

// Description: File that houses the package documentation for the cli package.

// Package cli contains various cli utilities that are useful for building
// cli applications with gobox.
package cli
```

File: pkg/cli/bootstrap.go (hunk 2 above)
Comment (jaredallard) [reply pkg/cli/bootstrap.go:48]: yes

**Review by george-e-shaw-iv — APPROVED** [review]

File: pkg/cli/bootstrap.go [hunk 6]
```diff
@@ -0,0 +1,262 @@
...
+	if err := trace.InitTracer(ctx, name); err != nil {
+		fmt.Println(err)
+		return ctx
+	}
+	return trace.StartTrace(ctx, name)
+}
+
+// setupPanicHandler sets up a panic handler for CLIs
+func setupPanicHandler(exitCode *int) {
```
Comment (vk-outreach) [inline pkg/cli/bootstrap.go:105]: lol.  I think we can make this a bit more readable by creating a temporary struct to hold all the state and moving all these top-level functions as methods on that struct..

i.e:

```golang

type cli struct {
    exitCode int
    cleanup func()
}

func (c *cli) setupPanicHandler() {
       if r := recover(); r != nil {
          c.exitCode = 2
       }
}
```

**Pull request merged**